README.md
# BoxCat

[![Language: Swift 3.0](https://img.shields.io/badge/lang-Swift_3.0-yellow.svg?style=flat)](https://developer.apple.com/swift/)
[![Platform](https://img.shields.io/cocoapods/p/BoxCat.svg?style=flat)](http://cocoapods.org/pods/BoxCat)
[![Version](https://img.shields.io/cocoapods/v/BoxCat.svg?style=flat)](http://cocoapods.org/pods/BoxCat)
[![License](https://img.shields.io/cocoapods/l/BoxCat.svg?style=flat)](http://cocoapods.org/pods/BoxCat)
[![CI Status](http://img.shields.io/travis/devdrey/BoxCat.svg?style=flat)](https://travis-ci.org/devdrey/BoxCat)
[![codecov](https://codecov.io/gh/devdrey/BoxCat/branch/develop/graph/badge.svg)](https://codecov.io/gh/devdrey/BoxCat)
[![DocCov](https://img.shields.io/cocoapods/metrics/doc-percent/BoxCat.svg)](http://cocoadocs.org/docsets/BoxCat)

*Bundled resource loading made easy*

_BoxCat_ provides an easy interface for loading resources from no matter were they are located in the application bundle. It's able to load resources from main app bundle as well from nested resource bundles and contained frameworks, e.g. Pods that containing images, xib-files and any other types of resources.

### Loading resources from Pods and resource bundles

Its main purpose is to load resources from resource bundles that may be contained in Pods. See [Cocoapods bundled resources](https://guides.cocoapods.org/syntax/podspec.html#resource_bundles) for more info. Thus for the unified interface it's also possible to exchange the iOS SDK solution for loading resources, especially if you plan to move some resources into pods in future. You won't even need to change code for loading the resources.

### Overriding resources

Using _BoxCat_ for loading resources will also enable the possibility to override resources contained in a pod with resources from main bundle. You simply need to include a resource in the main app bundle named the same as the one in the pod or resource bundle and loading it using the default interfaces of _BoxCat_ like `UIImage.named(_:)`.

## How to use

You using Box cat similar to the build in resource loading of the iOS SDK. But instead of using the resource classes initialiser to load the resource you use the interface _BoxCat_ provides.

To load an image resource you simply type.

```swift
let boxCatImage = UIImage.named("my_image")
```

This will search the whole main app bundle for the resource. So if its located in the main app bundle it will just load it from there. If it is contained in a bundle located inside of the main app bundle it will also find it and load it.

_BoxCat_ also provides `LookupFilter` to increase performance in searching for loading resources deeply contained in the main bundle hierarchy. Providing a concrete `Bundle` to the interfaces of _BoxCat_ is also a option to increase the performance loading resources.

You can specify those `LookupFilter` globally or locally only affecting a single resource loading call.

```swift
// Set lookup filter globally
// These lookup filter will be used for every resource loading call
let lookupFilter = lookupFilter(frameworks: ["GlobalResources"], 
                                bundles: ["Images"])
BoxCat.lookupFilter = lookupFilter
        
//...

// Set specific lookup filter at the resource loading call
// Overriding the globally set lookup filter
UIImage.named("bg_launch", lookupFilter: LookupFilter(frameworks: ["GlobalResources"], bundles: ["BigImages"]))

```

## Requirements

* iOS 8+
* Swift 3.0
* Cocoapods

## Installation

_BoxCat_ is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "BoxCat", :git => 'https://github.com/dreyhomedev/BoxCat.git'
```

## Example

The contained example shows how to load images from main bundle, a pod containing image resources directly and loading images from a pod containing a resource bundle. You will notice the call is always the same for each case!

You will also notice the use of the `LookupFilter` _BoxCat_ provides. They limit the searching to specific frameworks and (sub)bundles in the app bundle hierarchy. That enhances the loading speed of the resources if you have many pods included that my not contain resources like iOS Frameworks.

## Author

[@devdrey](devdrey@dreyhomenet.de)

## License

_BoxCat_ is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Changelog

## HEAD

## 1.0

* First release
* Loading image resources
* Loading nib resources


Example/Tests/Tests.swift
import UIKit
import XCTest
import BoxCat

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Example/Tests/LookupFilterTests.swift
//
//  LookupFilterTests.swift
//  BoxCat
//
//  Created by Andreas on 03.11.16.
//  Copyright © 2016 CocoaPods. All rights reserved.
//

import XCTest
@testable import BoxCat

class LookupFilterTests: XCTestCase {
    
    let mainBundleResource = "img_green"
    let frameworkResource = "img_red"
    let resourceBundleResource = "img_blue"
    
    var mainBundle: Bundle?
    
    override func setUp() {
        super.setUp()
        
        mainBundle = Bundle(for: LookupFilterTests.self)
    }
    
    override func tearDown() {
        mainBundle = nil
        BoxCat.lookupFilter = nil
        super.tearDown()
    }
    
    func testNoLookupFiter() {
        BoxCat.lookupFilter = nil
        
        XCTAssertNotNil(UIImage.named(mainBundleResource, bundle: mainBundle!))
        XCTAssertNil(UIImage.named(frameworkResource, bundle: mainBundle!))
        XCTAssertNil(UIImage.named(resourceBundleResource, bundle: mainBundle!))
    }
    
    func testAllBundlesAndFranworks() {
        BoxCat.lookupFilter = LookupFilter()
        
        XCTAssertNotNil(UIImage.named(mainBundleResource, bundle: mainBundle!))
        XCTAssertNotNil(UIImage.named(frameworkResource, bundle: mainBundle!))
        XCTAssertNotNil(UIImage.named(resourceBundleResource, bundle: mainBundle!))
        
        BoxCat.lookupFilter = LookupFilter(bundles: [], frameworks: [])
        
        XCTAssertNotNil(UIImage.named(mainBundleResource, bundle: mainBundle!))
        XCTAssertNotNil(UIImage.named(frameworkResource, bundle: mainBundle!))
        XCTAssertNotNil(UIImage.named(resourceBundleResource, bundle: mainBundle!))
        
    }
    
    func testBundlesOnlyLookupFiter() {
        BoxCat.lookupFilter = LookupFilter(bundles: [], frameworks: nil)
        
        XCTAssertNotNil(UIImage.named(mainBundleResource, bundle: mainBundle!))
        XCTAssertNil(UIImage.named(frameworkResource, bundle: mainBundle!))
        XCTAssertNil(UIImage.named(resourceBundleResource, bundle: mainBundle!))
    }
    
    func testFrameworksOnlyLookupFiter() {
        BoxCat.lookupFilter = LookupFilter(bundles: nil, frameworks: [])
        
        XCTAssertNotNil(UIImage.named(mainBundleResource, bundle: mainBundle!))
        XCTAssertNotNil(UIImage.named(frameworkResource, bundle: mainBundle!))
        XCTAssertNil(UIImage.named(resourceBundleResource, bundle: mainBundle!))
    }
    
    func testNoBundlesAndFrameworksLookupFiter() {
        BoxCat.lookupFilter = LookupFilter(bundles: nil, frameworks: nil)
        
        XCTAssertNotNil(UIImage.named(mainBundleResource, bundle: mainBundle!))
        XCTAssertNil(UIImage.named(frameworkResource, bundle: mainBundle!))
        XCTAssertNil(UIImage.named(resourceBundleResource, bundle: mainBundle!))
    }
    
    func testOverrideGlobalLookupFilter() {
        BoxCat.lookupFilter = nil
        
        XCTAssertNotNil(UIImage.named(mainBundleResource, bundle: mainBundle!))
        XCTAssertNotNil(UIImage.named(frameworkResource, bundle: mainBundle!, lookupFilter: LookupFilter()))
        XCTAssertNotNil(UIImage.named(resourceBundleResource, bundle: mainBundle!, lookupFilter: LookupFilter()))
    }
}


Example/ResourcePods/RedBoxCat/Example/Tests/Tests.swift
import UIKit
import XCTest
import RedBoxCat

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measureBlock() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


BoxCat/Classes/BoxCat.swift
//
//  BoxCat.swift
//  Pods
//
//  Created by Andreas on 01.10.16.
//
//

import Foundation

/// A filter to specify the destinations _BoxCat_ should look for resources.
/// It will also take into account the order of the defined destinations.
public struct LookupFilter {
    
    // The bundles to lookup for resources
    public let bundles: [String]?
    
    // The frameworks to lookup for resources
    public let frameworks: [String]?
    
    /// Creates a new `LookupFilter`
    ///
    /// - note: _BoxCat_ will return the first occurrence of a resource in the bundle hierarchy,
    ///         therfore it takes into account the order of the framewoks and bundles specified in the lookup 
    ///         filter.
    ///
    /// _BoxCat_ will search contained bundles before searching frameworks.
    ///
    /// - parameter bundles:    The names of the bundles to search for resources.
    ///                         If set to `nil` bundle lookup is skipped.
    ///                         If set to `[]` all contained bundles will be searched. _Default_
    /// - parameter frameworks: The names of the frameworks to search for resources.
    ///                         If set to `nil` framework lookup is skipped.
    ///                         If set to `[]` all contained frameworks will be searched. _Default_
    ///
    /// - returns: A new `LookupFilter`
    public init(bundles: [String]? = [], frameworks: [String]? = []) {
        self.frameworks = frameworks
        self.bundles = bundles
    }
}

public struct BoxCat {
    
    /// Global lookup filter. Defaults to an empty filter.
    /// If set to `nil` resource lookup in contained bundles and framewoks will be skipped.
    public static var lookupFilter: LookupFilter? = LookupFilter()
}

public extension UIImage {
    
    /// Searches for an image resource with a given name in the bundle structure and it's contained frameworks and bundles.
    ///
    /// - parameter name:         Name of the image resource.
    /// - parameter bundle:       A bundle to start the lookup. 
    ///                           If set to `nil` lookup will start at the main bundle. _Default_
    /// - parameter lookupFilter: A filter to narrow the lookup. _Default:_ `BoxCat.lookupFilter`
    ///
    /// - returns: Returns the first image resource found with `name` or `nil` if not found in the bundle hierarchy.
    public static func named(_ name: String, bundle: Bundle? = nil, lookupFilter: LookupFilter? = BoxCat.lookupFilter) -> UIImage? {
        return ResourceLoader.image(named: name, bundle: bundle, lookupFilter: lookupFilter)
    }
}

public extension UINib {
    
    /// Searches for a nib resource with a given name in the bundle structure and it's contained frameworks and bundles.
    ///
    /// - parameter name:         Name of the nib resource.
    /// - parameter bundle:       A bundle to start the lookup.
    ///                           If set to `nil` lookup will start at the main bundle. _Default_
    /// - parameter lookupFilter: A filter to narrow the lookup. _Default:_ `BoxCat.lookupFilter`
    ///
    /// - returns: Returns the first nib resource found with `name` or `nil` if not found in the bundle hierarchy.
    public static func named(_ name: String, bundle: Bundle? = nil, lookupFilter: LookupFilter? = BoxCat.lookupFilter) -> UINib? {
        return ResourceLoader.nib(named: name, bundle: bundle, lookupFilter: lookupFilter)
    }
}


Example/BoxCat/AppDelegate.swift
//
//  AppDelegate.swift
//  BoxCat
//
//  Created by Andreas Günther on 08/30/2016.
//  Copyright (c) 2016 Andreas Günther. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



BoxCat/Classes/ResourceLoader.swift
//
//  ResourceLoader.swift
//  Pods
//
//  Created by Andreas on 20.08.16.
//
//

import UIKit

class ResourceLoader: NSObject {

    static func image(named: String, bundle: Bundle? = nil, lookupFilter: LookupFilter?) -> UIImage? {
        
        let bundle = bundle ?? Bundle.main
        
        if let image = UIImage(named: named, in: bundle, compatibleWith: nil) {
            return image
        }
        
        return lookupResource(named: named, in: bundle, lookupFilter: lookupFilter) { (resourceName, subBundle) -> UIImage? in
            return ResourceLoader.image(named: resourceName, bundle: subBundle, lookupFilter: lookupFilter)
        }
    }
    
    static func nib(named: String, bundle: Bundle? = nil, lookupFilter: LookupFilter?) -> UINib? {
        
        let bundle = bundle ?? Bundle.main
        
        if bundle.path(forResource: named, ofType: "nib") != nil {
            return UINib(nibName: named, bundle: bundle)
        }
        
        return lookupResource(named: named, in: bundle, lookupFilter: lookupFilter, lookupClosure: { (resourceName, subBundle) -> UINib? in
            return ResourceLoader.nib(named: resourceName, bundle: subBundle, lookupFilter: lookupFilter)
        })
    }
    
    static func font(named: String, size: CGFloat) -> UIFont? {
        // load dynamically http://www.mokacoding.com/blog/sharing-assets-with-cocoapods-resource-bundle-and-dynamically-loaded-fonts/
        
        return nil
    }
    
    // MARK: private methods
    
    private static func lookupResource<ResourceType>(named: String, in bundle: Bundle, lookupFilter: LookupFilter?, lookupClosure: (_ resourceName: String, _ subBundle: Bundle) -> ResourceType?) -> ResourceType? {
        
        guard let lookupFilter = lookupFilter else {
            // no lookup filter specified. Ignoring contained bundles and frameworks.
            return nil
        }
        
        // lookup in contained bundles if the filter is specified.
        if let lookupBundles = lookupFilter.bundles {
            for subBundle in bundle.bundles.sortedIntersectionOfBundlesWith(names: lookupBundles) {
                if let resource = lookupClosure(named, subBundle) {
                    return resource
                }
            }
        }
        
        // lookup in contained frameworks if the filter is specified.
        if let lookupFrameworks = lookupFilter.frameworks {
            let frameworks = bundle.frameworks.sortedIntersectionOfBundlesWith(names: lookupFrameworks)
            for framework in frameworks {
                if let resource = lookupClosure(named, framework) {
                    return resource
                }
            }
        }
        
        return nil
    }
}


BoxCat/Classes/NSBundle+BoxCat.swift
//
//  NSBundle+BoxCat.swift
//  Pods
//
//  Created by Andreas on 20.08.16.
//
//

import Foundation

extension Bundle {
    
    var name: String {
        let bundleURL = URL(fileURLWithPath: self.bundlePath, isDirectory: true)
        let fileName = bundleURL.lastPathComponent
        let name = fileName.replacingOccurrences(of: ".\(bundleURL.pathExtension)", with: "")
        return name
    }
    
    var bundles: [Bundle] {
        return bundles("bundle")
    }
    
    var frameworks: [Bundle] {
        return bundles("framework", inDirectory: "Frameworks")
    }
    
    fileprivate func bundles(_ type: String, inDirectory directory: String? = nil) -> [Bundle] {
        let subBundlePaths = paths(forResourcesOfType: type, inDirectory: directory)
        return subBundlePaths.flatMap({Bundle(path: $0)})
    }
}

extension Array where Element: Bundle {
    
    func filteredBy(names: [String]) -> [Bundle] {
        if names.count == 0 {
            return self
        }
        return self.filter({ names.contains($0.name) })
    }
}

extension Array where Element: Bundle {
    
    func sortedIntersectionOfBundlesWith(names: [String]) -> [Bundle] {
        
        guard names.count > 0 else {
            return self
        }
        
        let intersection = self.filteredBy(names: names)
        let count = self.count > names.count ? self.count : names.count
        var sortedIntersection = Array<Bundle?>(repeating: nil, count: count)
        
        for unsortedItem in intersection {
            if let sortedIndex = names.index(of: unsortedItem.name) {
                sortedIntersection[sortedIndex] = unsortedItem
            }
        }
        return sortedIntersection.flatMap({ $0 })
    }
}


Example/BoxCat/ImageTableViewCell.swift
//
//  ImageTableViewCell.swift
//  BoxCat
//
//  Created by Andreas on 30.08.16.
//  Copyright © 2016 CocoaPods. All rights reserved.
//

import UIKit

class ImageTableViewCell: UITableViewCell {

    var bundleImage: UIImage? {
        get {
            return bundleImageView.image
        }
        set(newImage) {
            bundleImageView.image = newImage
        }
    }
    
    var bundleImageName: String? {
        get {
            return bundleImageNameLabel.text
        }
        set(newBundleImageName) {
            bundleImageNameLabel.text = newBundleImageName
        }
    }
    
    var bundleName: String? {
        get {
            return bundleNameLabel.text
        }
        set(newBundleName) {
            bundleNameLabel.text = newBundleName
        }
    }
    
    @IBOutlet fileprivate weak var bundleImageView: UIImageView!
    @IBOutlet fileprivate weak var bundleImageNameLabel: UILabel!
    @IBOutlet fileprivate weak var bundleNameLabel: UILabel!
}


Example/BoxCat/ImagesViewController.swift
//
//  ImagesViewController.swift
//  BoxCat
//
//  Created by Andreas Günther on 08/30/2016.
//  Copyright (c) 2016 Andreas Günther. All rights reserved.
//

import UIKit
import BoxCat

class ImagesViewController: UITableViewController {

    fileprivate let imageNames = ["img_green", "img_red", "img_blue"]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let lookupFilter = LookupFilter(bundles: ["BlueBoxCatResourcesBundle"],
                                        frameworks: ["BlueBoxCat", "RedBoxCat"])
        BoxCat.lookupFilter = lookupFilter
    }
    
    // MARK - UITableViewDataSource
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return imageNames.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let imageCell = tableView.dequeueReusableCell(withIdentifier: "ImageCell", for: indexPath) as! ImageTableViewCell
        
        guard let boxCatImage = UIImage.named(imageNames[(indexPath as NSIndexPath).row]) else {
            print("ImagesViewController > ERROR: loading image '\(imageNames[(indexPath as NSIndexPath).row])' failed!")
            return imageCell
        }
        
        imageCell.bundleImage = boxCatImage
        
        return imageCell
    }
}

